### PR TITLE
[LoopVectorize] Add an option to test potential issues with VPlan

### PR DESCRIPTION
**Pull request by danilaml**
Add a way to test for https://github.com/llvm/llvm-project/issues/115169.

### CODE REVIEW

**Comment by danilaml** [comment]
@fhahn what do you think? Using address spaces for that feels a bit cleaner but would also require bigger changes to the LoopVectorizer.cpp (or moving the option to Loads.cpp) which I'm less comfortable with.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -3448,8 +3453,9 @@ bool LoopVectorizationCostModel::isPredicatedInst(Instruction *I) const {
   // If predication is not needed, avoid it.
   // TODO: We can use the loop-preheader as context point here and get
   // context sensitive reasoning for isSafeToSpeculativelyExecute.
+  Instruction *CtxI = CheckSpeculatabilityAtUse ? I->getNextNode() : nullptr;
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3456]: Thinking more about this, wouldn't we need to ask if it is safe to execute speculatively in the loop header? Or a predecessors of `I->getParent()` that executes unconditionally?

Comment (danilaml) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3456]: If it was a refinement/TODO implementation, then yes. But this is an option strictly for testing to allow the problematic codegen so that it can be tested against. More "ideologically pure" alternatives would be less clean to implement (at least I haven't come up with a way to), like special-casing loads from specific address space (the AS value could be an option, but not sure how to name it).

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3456]: But in that case, wouldn't it be `isSafeToSpeculativelyExecute`'s responsibility to make sure that the pointer operand isn't poison?

Comment (danilaml) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3456]: It is easier to do in `isSafeToSpeculativelyExecute` but then it'd mean having a LoopVectorize-only testing option in it which doesn't sound like a good idea. Or did you have something else in mind? Basically, what I need to achieve to be able to test the codegen is to consider certain loads to be `isSafeToSpeculativelyExecute == true`.

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3456]: Hmm one thing I am missing is how `isSafeToSpeculativelyExecute == true` and the address being poison could happen. Even if the load is from a special address space or something like that, `isSafeToSpeculativelyExecute` should still consider the fact that the address could be poison?

Comment (danilaml) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3456]: @fhahn That's the thing. The address IS NOT poison initially and it CAN'T happen in original IR for which `isSafeToSpeculativelyExecute` is called (if I understood your question correctly). Load from poison only happens due to what I believe incorrect lowering of VPlan to IR (e.g. it does `select cond, (op x, poison), y` instead of `op x, (select cond, x, z)` or something else valid). I'd try to dig further, but unfortunately I couldn't make heads and tails of VPlan<>IR interop (downside of separate IR).

In short, `isSafeToSpeculativelyExecute == true` is fine. The address being poison is the bug introduced by LoopVectorize/VPlan. This patch (or alternative patches described in the linked issue) are an attempt to reproduce this issue on current main since after some changes I was unable to reproduce this issue with the upstream code alone (no way to specify always-deferenceable loads that are not args).

Comment (danilaml) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3456]: Ping. Any ideas on how to better trigger this scenario for testing?

Comment (danilaml) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3456]: ping
